Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+             nodeId_ = *(uint16_t *)pos;  // Save the node id.  Should be the same for all records.
+             pos += sizeof(uint16_t);
+             uint16_t sliceId = *(uint16_t *)pos;
+             pos += sizeof(uint16_t);
+             loss_ = *(float *)pos;  // Loss from previous
+             pos += sizeof(float);
+             minLoss_ = *(float *) pos; // Minimum observed loss
+             pos += sizeof(float);
+             minEpoch_ = *(uint32_t *) pos; // Epoch of minimum observed loss
+             pos += sizeof(uint32_t);
+             maxNoProg_ = *(uint32_t *) pos; // Maximum Epochs with no progress
+             pos += sizeof(uint32_t);
+             batchPos_ = *(uint64_t *) pos; // Current position in training set
+             pos += sizeof(uint64_t);
+             pos += sizeof(bool) + sizeof(uint32_t); // pass over SET's all flag and length
+             uint32_t offset = (sliceId-1) * slicesize * sizeof(double);
+             memcpy(((byte *)fW_ + offset), pos, slicesize * sizeof(double));
+             rtlReleaseRow(next);
+          }
+          // Get the training records and store in flattened training array (fT_).
+          uint32_t i = 0, j = 0;
+          while (true)
+          {
+            const byte * next = (const byte *)train->nextRow();
+            if (!next) break;
+            byte * pos = (byte *) next;
+            // Add to flattened training set
+            fT_[fT_count_].main = *(uint32_t *) pos; // Main word
+            pos += sizeof(uint32_t);
+            // Set of context words.  Bypass 'ALL' flag.
+            pos += sizeof(bool);
+            // Calculate the number of context words.
+            uint32_t contxLen = *(uint32_t *) pos;
+            uint32_t contxCount = contxLen / sizeof(uint32_t);
+            fT_[fT_count_].contxCount = contxCount;
+            // Allocate storage for the context array
+            uint32_t * context = (uint32_t *)rtlMalloc(contxLen);
+            fT_[fT_count_].contx = context;
+            pos += sizeof(uint32_t);
+            // Copy the context wordIds into the array
+            memcpy((byte *)context, pos, contxLen);
+            rtlReleaseRow(next);
+            fT_count_++;
+          }
+          // Now we have the weights and the training pairs.  Create a neural network
+          // and train it on this data.
+          nn_ = new NeurNet(fW_, fU_, numwords, dim, lr, negsamp);
+          nn_->Train(fT_, fT_count_);
+          // Calculate the overall loss for this batch
+          loss_ = nn_->getLoss();
+          // Delete the neural network.  We're done with it.
+          delete nn_;
+          // Mark so we know we've already calculated the results.
+          calculated_ = true;
+        } // end if (!calculated)
+        // Now return the next slice.
+        // Copy the weight updates into slices to return.
+        uint32_t returnSize;
+        byte * row;
+        RtlDynamicRowBuilder rowBuilder(resultAllocator);
```

Comment:
trivial: move inside the if() true branch

Reply:
deferred